src/metabase/lib/field.cljc

```diff
@@ -0,0 +1,90 @@
+(ns metabase.lib.field
+  (:require
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.schema.ref :as lib.schema.ref]
+   [metabase.lib.temporal-bucket :as lib.temporal-bucket]
+   [metabase.lib.util :as lib.util]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(defmulti ^:private ->field
+  {:arglists '([query stage-number field])}
+  (fn [_query _stage-number field]
+    (lib.dispatch/dispatch-value field)))
+
+(defmethod ->field :metadata/field
+  [_query _stage-number field-metadata]
+  (lib.options/ensure-uuid
+   (or (:field_ref field-metadata)
+       (when-let [id (:id field-metadata)]
+         [:field id nil])
+       [:field (:name field-metadata) {:base-type (:base_type field-metadata)}])))
+
+(defn- resolve-field [query stage-number table-or-nil id-or-name]
+  (let [metadata (letfn [(field-metadata [metadata]
+                           (when metadata
+                             (lib.metadata/field-metadata metadata table-or-nil id-or-name)))]
+                   ;;; TODO -- I think maybe this logic should actually be part of [[metabase.lib.metadata]]
+                   ;;; -- [[metabase.lib.field-metadata]] should be the one doing this nonsense
+                   (or
+                    ;; resolve field from the current stage
+                    (field-metadata (lib.metadata/stage-metadata query stage-number))
+                    ;; resolve field from the previous stage (if one exists)
+                    (when (lib.util/has-stage? query (dec stage-number))
+                      (field-metadata (lib.metadata/stage-metadata query (dec stage-number))))
+                    ;; resolve field from Database metadata
+                    (field-metadata (lib.metadata/database-metadata query))))]
```

Comment:
yup, I agree. I honestly don't have this all figured out 100% yet. I'm trying to improve the way metadata works in #28717, but I'm not convinced that approach is perfect either -- it will probably take a few PRs to really get this nailed down perfectly. But we don't need to get everything perfect right away